### ISSUE TITLE
Generated POM or Ivy descriptor for replaced module doesn't reflect proper dependencies

### ISSUE THREAD

**Issue by rpalcolea** [gradle/gradle#9180]
### Expected Behavior
When a direct dependency is replaced via module replacement, maven POM and Ivy descriptor should use the replacement as part of the final XML.

### Current Behavior
Currently, it writes the original dependency.

### Steps to Reproduce 
This is a sample project for reproducing this: https://github.com/nebula-plugins/gradle-nebula-integration/tree/master/gradle-module-replacement-test

Given this build script:

```
import gradle.module.replacement.test.plugin.DependencyPlugin

plugins {
    // Apply the java-library plugin to add support for Java Library
    id 'java-library'
    id 'maven-publish'
    id 'ivy-publish'

}

apply plugin: DependencyPlugin

repositories {
    jcenter()
}

dependencies {
     api 'com.google.truth:truth:0.28'
     modules {
         module('com.google.collections:google-collections') {
             replacedBy('com.google.guava:guava')
         }
     }
}

publishing {
    publications {
        maven(MavenPublication) {
            from components.java
            versionMapping {
                allVariants{
                    fromResolutionResult()
                }
            }

        }
        ivy(IvyPublication) {
            from components.java
            versionMapping {
                allVariants{
                    fromResolutionResult()
                }
            }

        }
    }
}
```

We are replacing google-collections with guava. If we introduce a plugin that adds dependencies (similar to what can be found in other plugins such as lombok), for example:

```
class DependencyPlugin implements Plugin<Project> {
    @Override
    void apply(Project project) {
        addCollectionsDependendency(project, "api")
    }

    private void addCollectionsDependendency(Project project, String configurationName) {
        project.dependencies.add(
                configurationName,
                "com.google.collections:google-collections:1.0"
        )
    }
}
```

When this plugin is applied, I can see the following dependency graph:

```
compileClasspath - Compile classpath for source set 'main'.
+--- com.google.collections:google-collections:1.0 -> com.google.guava:guava:18.0
\--- com.google.truth:truth:0.28
     +--- com.google.guava:guava:18.0
     \--- junit:junit:4.10
          \--- org.hamcrest:hamcrest-core:1.1
```

Replacement happens, however... when publishing a POM or Ivy descriptor we get the following:

```
?xml version="1.0" encoding="UTF-8"?>
<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="http://maven.apache.org/POM/4.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <modelVersion>4.0.0</modelVersion>
  <groupId></groupId>
  <artifactId>gradle-module-replacement-test</artifactId>
  <version>unspecified</version>
  <dependencies>
    <dependency>
      <groupId>com.google.collections</groupId>
      <artifactId>google-collections</artifactId>
      <version>1.0</version>
      <scope>compile</scope>
    </dependency>
    <dependency>
      <groupId>com.google.truth</groupId>
      <artifactId>truth</artifactId>
      <version>0.28</version>
      <scope>compile</scope>
    </dependency>
  </dependencies>
</project>
``` 

and module info:

```
{
  "formatVersion": "1.0",
  "component": {
    "group": "",
    "module": "gradle-module-replacement-test",
    "version": "unspecified",
    "attributes": {
      "org.gradle.status": "integration"
    }
  },
  "createdBy": {
    "gradle": {
      "version": "5.4",
      "buildId": "ifwmcnxslvewrcmeypdqqhpjja"
    }
  },
  "variants": [
    {
      "name": "apiElements",
      "attributes": {
        "org.gradle.category": "library",
        "org.gradle.dependency.bundling": "external",
        "org.gradle.jvm.version": 8,
        "org.gradle.usage": "java-api-jars"
      },
      "dependencies": [
        {
          "group": "com.google.collections",
          "module": "google-collections",
          "version": {
            "requires": "1.0"
          }
        },
        {
          "group": "com.google.truth",
          "module": "truth",
          "version": {
            "requires": "0.28"
          }
        }
      ],
      "files": [
        {
          "name": "gradle-module-replacement-test-unspecified.jar",
          "url": "gradle-module-replacement-test-unspecified.jar",
          "size": 1147,
          "sha1": "287897e93d34ddabe79418773ad9586e0daf81ab",
          "md5": "db4e7a3119fa94e1c5d7e8094dd9aa0e"
        }
      ]
    },
    {
      "name": "runtimeElements",
      "attributes": {
        "org.gradle.category": "library",
        "org.gradle.dependency.bundling": "external",
        "org.gradle.jvm.version": 8,
        "org.gradle.usage": "java-runtime-jars"
      },
      "dependencies": [
        {
          "group": "com.google.collections",
          "module": "google-collections",
          "version": {
            "requires": "1.0"
          }
        },
        {
          "group": "com.google.truth",
          "module": "truth",
          "version": {
            "requires": "0.28"
          }
        }
      ],
      "files": [
        {
          "name": "gradle-module-replacement-test-unspecified.jar",
          "url": "gradle-module-replacement-test-unspecified.jar",
          "size": 1147,
          "sha1": "287897e93d34ddabe79418773ad9586e0daf81ab",
          "md5": "db4e7a3119fa94e1c5d7e8094dd9aa0e"
        }
      ]
    }
  ]
}
``` 

We would expect the guava replacement to be reflected in the POM / module files.

**Comment by dsilvasc** [comment]
Is this related to https://github.com/gradle/gradle/issues/9565 ?

**Comment by ljacomet** [comment]
No, it is different, although the end solution is similar.

So effectively this was fixed through #10019. Note that the substitution _only takes place_ when publishing resolved versions.
Without that behavior, Gradle does not even attempt resolution and thus has no way to know that a first level dependency was substituted.
